### ISSUE TITLE
[Request] Support for video types other than Theora?

### ISSUE THREAD

**Issue by SirAxolot** [love2d/love#2089]
I was wondering if it would be possible to support videos other than Ogg Theora in LOVE2D?
The one way that comes to mind on how to do this is using [libVLC](https://www.videolan.org/vlc/libvlc.html), specifically [libvlcpp](https://code.videolan.org/videolan/libvlcpp), the C++ version of it. Since libVLC is what's used for VLC Media player, theoretically it would be able to support ALL video and audio types.
As I don't really know a lot of cpp im not sure how hard this would be to implement but if it would be relatively easy then it would be realy cool if i could just directly use mp4 files rather than converting to Theora

**Comment by MikuAuahDark** [comment]
Personal opinion: Ogg Theora sucks. The reference decoder is too slow even FFmpeg's own implementation beats them.

I have separate LOVE branch containing integration with my WIP cross-platform audio and video decoder called [NAV](https://github.com/MikuAuahDark/nav) which can be found here: https://github.com/MikuAuahDark/love2d/tree/nav. So far both audio and video integration is promising. If it gets accepted, then we may use that to provide wider audio.video support in LOVE.

So if you'd like to contribute using libVLC, I recommend doing so in NAV.

**Comment by SirAxolot** [comment]
> So if you'd like to contribute using libVLC, I recommend doing so in NAV.

Well i wouldnt be the one contributing this i dont know really anything about c++ LOL

i was just making a suggestion as an alternative to theora since its so slow and bloated

**Comment by andryeltj** [comment]
The mpv from libmpv have this lib to embed to others projects, and is writted in C. Like i dont know about implement this, here is my request for anyone who can

**Comment by MikuAuahDark** [comment]
I don't think we can use mpv (or specifically libmpv) due to licensing issue. We generally in favor for libraries with more permissive license compared to (L)GPL because we can't ship those libraries (which means a feature that depend on those copyleft license) in iOS.

**Comment by axunes** [comment]
might I suggest [pl_mpeg](https://github.com/phoboslab/pl_mpeg)?

**Comment by MikuAuahDark** [comment]
Unfortunately MPEG-1 has even worse quality than Theora, so no.

**Comment by axunes** [comment]
mpeg-1 quality's bad but it should perform a lot better, right? or mpeg-2?
if you can find a good software decoder implementation then H.264 would be awesome since it can be hardware-decoded on pretty much everything

**Comment by MikuAuahDark** [comment]
Theora is competing with MPEG-4 Part 2, so Theora also has better quality than MPEG-2. That being said, using Theora video for 720p60 playback in LOVE is fully possible. At 1080p60 you'll start to see stuttering.

Unfortunately we also cannot use any form of H264/H265 libraries because the algorithm behind that video codec is patented.

**Comment by ancientstraits** [comment]
I think I could help with this, specifically with using FFmpeg and integrating its C libraries into Love2D. I have already written a [project](https://github.com/ancientstraits/codim) that writes videos using FFmpeg. However, I don't know if this would work with not being able to use H264 or H265.

**Comment by MikuAuahDark** [comment]
NAV that I mentioned pretty much almost solved the problem. Unlike FFmpeg that's larger than whole LOVE itself, NAV is small enough to integrate and supports many backends (OS APIs such as Windows MediaFoundation and 3rd-party such as FFmpeg).

Still, I think the missing feature of that library right now is Apple platform support, more testing, and HW-accelerated video. Videos can be decoded using HW-acceleration in MediaFoundation right now but currently frame retrieval needs to move the video frame to system RAM. Unfortunately, it hurts performance alot that performant 4k60 playback is not possible with current approach.

**Comment by ancientstraits** [comment]
Okay, I will try helping with NAV then
